### DISCUSSION TITLE
Reddit gallery links

### DISCUSSION THREAD

**Discussion by JarheadHME** [tophf/mpiv#124] [MPIV host rules]
```json
{
 "r": "reddit\\.com/gallery/.+",
 "g": "return Array.from(doc.querySelectorAll(\"a.iUP9nbvcaxfwKrQTgt0sw\")).map(_ => ({url:\"https://i.redd.it/\" + _.href?.match(/.*redd\\.it\\/(\\w+\\.\\w+).*/)?.[1]}))"
}
```

I believe this works for all reddit galleries, as according to the uploader they can only have images and gifs. the only minor complication is images use `preview.redd.it/(filename)?(extra stuff)` and the gifs just do direct `i.redd.it/(filename)`, but the regex works all the same to correct both to `i.redd.it`

I took a quick peak at making one for reddit videos (`v.redd.it` links) and decided it was out of my wheelhouse, dash player stuff. If you know anything about that, you could maybe try to find how [RES](https://github.com/honestbleeps/Reddit-Enhancement-Suite/) does it

Honestly, I'm a little confused as to why just 
```json
{
 "r": "reddit\\.com/gallery/.+",
 "g": {
  "entry": "a.iUP9nbvcaxfwKrQTgt0sw"
 }
}
 ```
... doesn't just work. It grabs the first image, but none of the extras. (You can test this on this post: https://old.reddit.com/r/TheOdysseyHadAPurpose/comments/1h4vdgj/fishmael_toeiki_c/) I find the docs around "g" to be a bit, lacking.

**Comment by tophf** [comment]
Looking at the initial html in devtools network panel when opening the gallery link I see that the response doesn't have images inside those `a` elements (the site adds them on demand), so the simple gallery parser can't them.

This works for me:
```js
{"u": "||reddit.com/gallery/", "g": {"image": "figure a"}}
```

1. "entry" is necessary when you also want to find a title of each image, otherwise use "image".
2. Instead of the volatile identifiers like iUP9nbvcaxfwKrQTgt0sw that change each time the site is updated, it's better to find a stable way to locate the element using its attributes or parents.

**Comment by JarheadHME** [comment]
I clearly still have a lot to learn about how this stuff works, especially that just "figure a" works
I presume cause it's an `a` tag under a `figure` tag, but i did not expect it to, nor see that it could, work like that. My basic-hobbyist is showing haha

Well regardless, here's that but with a fix function that essentially just, cleans the image preview urls into the `i.redd.it` urls as the original submission did:

```json
{"u": "||reddit.com/gallery/",
 "g": {
  "image": "figure a",
  "fix": "return isURL ? (result = s.match(/.*redd\\.it\\/(\\w+\\.\\w+).*/))?.[1] ? \"https://i.redd.it/\" + result[1] : s : s"
 }
}
```

It did seem to work fine without it, though I kinda *assume* that all the extraneous bits at the end of the `preview.redd.it` urls might make it take a little bit longer? if you know that to be false, or at least not any worse than just running the fix then feel free to just axe it. Thanks though!

**Reply by tophf** [reply]
Nice. Can I add it to the rule gallery?

**Reply by JarheadHME** [reply]
Yep, feel free to!

**Comment by mechasharkbait** [comment]
The code in the rule gallery was working great until this morning and now I'm getting "Empty gallery" for every gallery on reddit.

**Reply by JarheadHME** [reply]
Seems reddit changed it from directing to a gallery page itself to the comments. Fairly simple fix in the end though:

```json
{"u": "||reddit.com/gallery/",
 "g": {
  "image": "div.gallery-tiles img.preview",
  "fix": "return isURL ? (result = s.match(/.*redd\\.it\\/(\\w+\\.\\w+).*/))?.[1] ? \"https://i.redd.it/\" + result[1] : s : s"
 }
}
```

Just doing `div.gallery-tiles img` can result in duplicated images due to there being two of each in nsfw posts, so it's `img.preview` instead.

@tophf verify and update pls 🙏

**Comment by krisu5** [comment]
None of the solutions in this thread are working for me and my low knowledge of programming couldn't figure out a solution 😅

**Reply by tophf** [reply]
Try the `Find rules` button in MPIV config, see https://github.com/tophf/mpiv/blob/master/README.md

**Reply by krisu5** [reply]
I have tried those too before trying these (namely: "old.reddit gallery in post (hover the title)" and "reddit upload image links"), doesn't work either.

For example, hovering the title [in this post](https://old.reddit.com/r/unixporn/comments/1mhcznw/xfce_a_little_bit_of_nostalgia_trip/) shows error cursor and "empty gallery" caption. Same with hovering over "reddit.com/gallery" links on [search page](https://old.reddit.com/r/unixporn/search?q=url%3Ahttps%3A%2F%2Fwww.reddit.com%2Fgallery&restrict_sr=on&sort=new&t=all).

Seperated images do work, even without rules. It's possible though that some [Betterfox user.js](https://github.com/yokoffing/Betterfox) rule does effect on this, I already tried with Violentmonkey and MPIV userscript only. EDIT: tested with fresh Firefox profile (without any about:config stuff and only Violentmonkey/MPIV), now I'm 100% sure that rules just don't work.
